code/datums/components/acid.dm

```diff
@@ -0,0 +1,178 @@
+/** Component representing acid applied to an object.
+  *
+  * Must be attached to an atom.
+  * Processes, repeatedly damaging whatever it is attached to.
+  * If the parent atom is a turf it applies acid to the contents of the turf.
+  */
+/datum/component/acid
+	dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS
+	/// The strength of the acid on the parent object.
+	var/acid_power
+	/// The volume of acid on the parent object.
+	var/acid_volume
+	/// The ambiant sound of acid eating away at the parent object.
+	var/datum/looping_sound/sizzle
+	/// Used exclusively for melting turfs. TODO: Move integrity to the atom level so that this can be dealt with there.
+	var/parent_integrity = 30
+
+/datum/component/acid/Initialize(_acid_power, _acid_volume)
+	if((_acid_power) <= 0 || (_acid_volume <= 0))
+		return COMPONENT_INCOMPATIBLE // Not enough acid.
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE // Incompatible type. TODO: Rework take_damage to the atom level and move this there.
+	if(isobj(parent))
+		var/obj/O = parent
+		if(O.resistance_flags & UNACIDABLE) // The parent object cannot have acid.
+			return COMPONENT_INCOMPATIBLE
+
+	acid_power = _acid_power
+	set_volume(_acid_volume)
+
+	var/atom/parent_atom = parent
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS, .proc/on_update_overlays)
+	parent_atom.update_icon()
+	sizzle = new/datum/looping_sound/acid(list(parent), TRUE)
+	START_PROCESSING(SSacid, src)
+
+/datum/component/acid/Destroy(force, silent)
+	STOP_PROCESSING(SSacid, src)
+	qdel(sizzle)
+	UnregisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS)
+	if(parent && !QDELETED(parent))
+		var/atom/parent_atom = parent
+		parent_atom.update_icon()
+	return ..()
+
+/datum/component/acid/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/on_examine)
+	RegisterSignal(parent, COMSIG_COMPONENT_CLEAN_ACT, .proc/on_clean)
+	RegisterSignal(parent, COMSIG_ATOM_ATTACK_HAND, .proc/on_attack_hand)
+	if(isturf(parent))
+		RegisterSignal(parent, COMSIG_MOVABLE_CROSSED, .proc/on_crossed)
+
+/datum/component/acid/UnregisterFromParent()
+	UnregisterSignal(parent, list(	\
+		COMSIG_PARENT_EXAMINE,		\
+		COMSIG_COMPONENT_CLEAN_ACT,	\
+		COMSIG_ATOM_ATTACK_HAND		))
+
+	if(isturf(parent))
+		UnregisterSignal(parent, COMSIG_MOVABLE_CROSSED)
+
+/// Averages corrosive power and sums volume.
+/datum/component/acid/InheritComponent(datum/component/C, i_am_original, _acid_power, _acid_volume)
+	acid_power = ((acid_power * acid_volume) + (_acid_power * _acid_volume)) / (acid_volume + _acid_volume)
+	set_volume(acid_volume + _acid_volume)
+
+/// Sets the acid volume to a new value. Limits the acid volume by the amount allowed to exist on the parent atom.
+/datum/component/acid/proc/set_volume(new_volume)
+	if(isobj(parent))
+		new_volume = clamp(new_volume, 0, OBJ_ACID_VOLUME_MAX)
+	else if(isturf(parent))
```

Comment:
Instead of checking types each time, store the difference of behaviors in variables, and only check for type on init.